ecs-agent/acs/session/task_manifest_responder_test.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	StartingTaskManifestSequenceNumber = 15
+	EndingTaskManifestSequenceNumber   = 20
+)
+
+// setupTestManifestMessage initializes a dummy TaskManifestMessage for testing
+func setupTestManifestMessage() *ecsacs.TaskManifestMessage {
+	return &ecsacs.TaskManifestMessage{
+		ClusterArn:           aws.String(testconst.ClusterName),
+		ContainerInstanceArn: aws.String(testconst.ContainerInstanceARN),
+		MessageId:            aws.String(testconst.MessageID),
+		Tasks:                []*ecsacs.TaskIdentifier{},
+		Timeline:             aws.Int64(StartingTaskManifestSequenceNumber),
+	}
+}
+
+// TestManifestAckHappyPath tests the happy path for a typical TaskManifestMessage and confirms
+// expected ack is made
+func TestManifestAckHappyPath(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	testManifest := setupTestManifestMessage()
+
+	stopVerifSent := make(chan *ecsacs.TaskStopVerificationMessage)
+
+	testResponseSender := func(response interface{}) error {
+		ackRequest, isAckRequest := response.(*ecsacs.AckRequest)
+		if isAckRequest {
+			// Validate ack request fields when happy path is reached.
+			require.Equal(t, aws.StringValue(ackRequest.MessageId), testconst.MessageID)
+		} else {
+			stopVerification, isStopVerification := response.(*ecsacs.TaskStopVerificationMessage)
+			if isStopVerification {
+				stopVerifSent <- stopVerification
+			}
+		}
+		return nil
+	}
+
+	mockMetrics := mock_metrics.NewMockEntryFactory(ctrl)
+	mockComparer := mock_session.NewMockTaskComparer(ctrl)
+	mockIDA := mock_session.NewMockManifestMessageIDAccessor(ctrl)
+	mockSNA := mock_session.NewMockSequenceNumberAccessor(ctrl)
+
+	testTaskManifestResponder := NewTaskManifestResponder(mockComparer, mockSNA, mockIDA, mockMetrics, testResponseSender)
+
+	// Expect to get the latest sequence number for stale check, then set it as we expect.
+	mockSNA.EXPECT().GetLatestSequenceNumber()
+	mockSNA.EXPECT().SetLatestSequenceNumber(testManifest).Return(nil)
+
+	// After we get the latest sequence number, we will expect to set the latest message ID.
+	mockIDA.EXPECT().SetMessageID(testconst.MessageID).Return(nil)
+
+	// Test Metric publication when our responder starts and finishes message processing.
+	mockEntry := mock_metrics.NewMockEntry(ctrl)
+	mockMetrics.EXPECT().New(metrics.TaskManifestHandlingDuration).Return(mockEntry)
+	mockEntry.EXPECT().Done(nil).AnyTimes()
+
+	// When the client queries tasks to be stopped, inject this task to be stopped.
+	mockComparer.EXPECT().CompareRunningTasksOnInstanceWithManifest(testManifest).Return(
+		[]*ecsacs.TaskIdentifier{
+			{DesiredStatus: aws.String("RUNNING"), TaskArn: aws.String(testconst.TaskARN)},
+		},
+		nil,
+	)
+
+	// Handle the task manifest message update.
+	handleManifestMessage := testTaskManifestResponder.HandlerFunc().(func(*ecsacs.TaskManifestMessage))
+	go handleManifestMessage(testManifest)
```

Comment:
This asynchronous call seems unnecessary. I think we should be able to assert the right behavior without needing channels.

Reply:
This is the paradigm we have been using in the agent and acs/session. It is minimal overhead so I think we should stick with this behavior.

Reply:
Synced offline, now validating expected data in the testrespondfunc, which should trigger inside of the responder rather than waiting until the end of test.